File: springboot/src/main/java/com/freshmall/service/CartService.java
package com.freshmall.service;

import com.freshmall.common.AppException;
import com.freshmall.entity.Goods;
import com.freshmall.repository.GoodsRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.redis.core.StringRedisTemplate;
import org.springframework.stereotype.Service;

import java.util.*;

/**
 * 购物车服务（对应 Node 版 services/cartService.js）
 * 存储：Redis Hash，key = cart:{userId}，field = goods_id，value = quantity
 */
@Service
@RequiredArgsConstructor
public class CartService {

    private final StringRedisTemplate redisTemplate;
    private final GoodsRepository goodsRepository;

    private String cartKey(Integer userId) {
        return "cart:" + userId;
    }

    private Goods checkGoods(Integer goodsId) {
        Goods goods = goodsRepository.findById(goodsId)
                .filter(Goods::getStatus)
                .orElseThrow(() -> new AppException("商品不存在或已下架", 404, 404));
        return goods;
    }

    /** 添加商品（数量累加） */
    public int addToCart(Integer userId, Integer goodsId, int quantity) {
        Goods goods = checkGoods(goodsId);
        String key = cartKey(userId);
        String field = String.valueOf(goodsId);

        String currentStr = (String) redisTemplate.opsForHash().get(key, field);
        int current = currentStr == null ? 0 : Integer.parseInt(currentStr);
        int total = current + quantity;
        if (total > goods.getStock()) {
            throw new AppException("商品库存不足，当前库存 " + goods.getStock() + goods.getUnit(), 400, 400);
        }
        redisTemplate.opsForHash().increment(key, field, quantity);
        return total;
    }

    /** 修改数量（覆盖） */
    public void updateQuantity(Integer userId, Integer goodsId, int quantity) {
        if (quantity <= 0) {
            throw new AppException("数量必须大于 0", 400, 400);
        }
        Goods goods = checkGoods(goodsId);
        if (quantity > goods.getStock()) {
            throw new AppException("商品库存不足，当前库存 " + goods.getStock() + goods.getUnit(), 400, 400);
        }
        redisTemplate.opsForHash().put(cartKey(userId), String.valueOf(goodsId), String.valueOf(quantity));
    }

    /** 删除商品 */
    public void removeFromCart(Integer userId, Integer goodsId) {
        redisTemplate.opsForHash().delete(cartKey(userId), String.valueOf(goodsId));
    }

    /**
     * 获取购物车（关联商品信息 + 小计）
     * 返回 [{ goods_id, quantity, goods, subtotal }]
     */
    public List<Map<String, Object>> getCart(Integer userId) {
        Map<Object, Object> raw = redisTemplate.opsForHash().entries(cartKey(userId));
        if (raw.isEmpty()) return List.of();

        List<Integer> ids = new ArrayList<>();
        Map<Integer, Integer> quantityMap = new HashMap<>();
        raw.forEach((field, value) -> {
            Integer goodsId = Integer.parseInt((String) field);
            ids.add(goodsId);
            quantityMap.put(goodsId, Integer.parseInt((String) value));
        });

        // 批量查商品
        List<Goods> goodsList = goodsRepository.findAllById(ids);
        Map<Integer, Goods> goodsMap = new HashMap<>();
        goodsList.forEach(g -> goodsMap.put(g.getId(), g));

        List<Map<String, Object>> result = new ArrayList<>();
        ids.stream().distinct().forEach(goodsId -> {
            Goods goods = goodsMap.get(goodsId);
            if (goods == null) return; // 过滤已删除/下架商品
            Map<String, Object> item = new LinkedHashMap<>();
            item.put("goods_id", goodsId);
            item.put("quantity", quantityMap.get(goodsId));
            item.put("goods", toGoodsMap(goods));
            item.put("subtotal", goods.getPrice().multiply(java.math.BigDecimal.valueOf(quantityMap.get(goodsId))));
            result.add(item);
        });
        return result;
    }

    /** 下单后清除指定商品 */
    public void clearCartItems(Integer userId, List<Integer> goodsIds) {
        if (goodsIds == null || goodsIds.isEmpty()) return;
        redisTemplate.opsForHash().delete(cartKey(userId), goodsIds.stream().map(String::valueOf).toArray());
    }

    /** 商品列表视图（与 Node 版一致，不含大字段） */
    private Map<String, Object> toGoodsMap(Goods g) {
        Map<String, Object> m = new LinkedHashMap<>();
        m.put("id", g.getId());
        m.put("category_id", g.getCategory_id());
        m.put("name", g.getName());
        m.put("image", g.getImage());
        m.put("price", g.getPrice());
        m.put("original_price", g.getOriginal_price());
        m.put("unit", g.getUnit());
        m.put("stock", g.getStock());
        m.put("sales", g.getSales());
        m.put("is_hot", g.getIs_hot());
        m.put("is_new", g.getIs_new());
        return m;
    }
}

File: server/services/cartService.js
/**
 * 购物车服务层
 * 存储：Redis Hash，key = cart:{userId}，field = goods_id，value = quantity
 * 结构：{ goodsId: quantity }
 */
const { Op } = require('sequelize')
const redis = require('../utils/redis')
const { Goods } = require('../models')
const AppError = require('../utils/appError')

const cartKey = (userId) => `cart:${userId}`

// 校验商品是否存在且上架
async function checkGoods(goodsId) {
  const goods = await Goods.findOne({ where: { id: goodsId, status: true } })
  if (!goods) {
    throw new AppError('商品不存在或已下架', 404, 404)
  }
  return goods
}

/**
 * 添加商品到购物车（数量累加）
 * @returns {number} 该商品在购物车中的最新数量
 */
async function addToCart(userId, goodsId, quantity) {
  const goods = await checkGoods(goodsId)
  const key = cartKey(userId)

  const current = parseInt(await redis.hGet(key, String(goodsId))) || 0
  const total = current + quantity
  if (total > goods.stock) {
    throw new AppError(`商品库存不足，当前库存 ${goods.stock}${goods.unit}`, 400, 400)
  }

  await redis.hIncrBy(key, String(goodsId), quantity)
  return total
}

/**
 * 修改购物车中某商品数量（直接覆盖）
 */
async function updateQuantity(userId, goodsId, quantity) {
  if (quantity <= 0) {
    throw new AppError('数量必须大于 0', 400, 400)
  }
  const goods = await checkGoods(goodsId)
  if (quantity > goods.stock) {
    throw new AppError(`商品库存不足，当前库存 ${goods.stock}${goods.unit}`, 400, 400)
  }

  await redis.hSet(cartKey(userId), String(goodsId), quantity)
}

/**
 * 删除购物车中某商品
 */
async function removeFromCart(userId, goodsId) {
  await redis.hDel(cartKey(userId), String(goodsId))
}

/**
 * 获取购物车（关联商品信息，含小计）
 * @returns {Array<{goods_id, quantity, goods, subtotal}>}
 */
async function getCart(userId) {
  const raw = await redis.hGetAll(cartKey(userId))
  const entries = Object.entries(raw).map(([goodsId, quantity]) => ({
    goods_id: parseInt(goodsId),
    quantity: parseInt(quantity),
  }))
  if (!entries.length) return []

  // 批量查询商品，合并为 { 购物车条目, 商品信息, 小计 }
  const ids = entries.map((e) => e.goods_id)
  const goodsList = await Goods.findAll({
    where: { id: { [Op.in]: ids } },
    attributes: { exclude: ['images', 'detail'] },
  })
  const goodsMap = new Map(goodsList.map((g) => [g.id, g]))

  return entries
    .filter((e) => goodsMap.has(e.goods_id)) // 过滤已删除/下架商品
    .map((e) => {
      const goods = goodsMap.get(e.goods_id)
      return {
        ...e,
        goods,
        subtotal: parseFloat(goods.price) * e.quantity,
      }
    })
}

/**
 * 清空购物车（下单成功后调用）
 */
async function clearCart(userId) {
  await redis.del(cartKey(userId))
}

/**
 * 删除购物车中指定商品（下单后清除已下单商品）
 * @param {number} userId
 * @param {number[]} goodsIds
 */
async function clearCartItems(userId, goodsIds) {
  if (!goodsIds || !goodsIds.length) return
  await redis.hDel(cartKey(userId), goodsIds.map(String))
}

module.exports = { addToCart, updateQuantity, removeFromCart, getCart, clearCart, clearCartItems }
